pubnub/event_engine/effects.py

```diff
@@ -0,0 +1,431 @@
+
+        if feature_enabled('PN_MAINTAIN_PRESENCE_STATE') and hasattr(self.pubnub, 'state_container'):
+            state_container = self.pubnub.state_container
+            request.state(state_container.get_state(self.invocation.channels))
+
+        response = await request.future()
+
+        if isinstance(response, PubNubException):
+            self.logger.warning(f'Heartbeat failed: {str(response)}')
+            self.event_engine.trigger(events.HeartbeatFailureEvent(channels=channels, groups=groups,
+                                                                   reason=response.status.error_data, attempt=1))
+        elif response.status.error:
+            self.logger.warning(f'Heartbeat failed: {response.status.error_data.__dict__}')
+            self.event_engine.trigger(events.HeartbeatFailureEvent(channels=channels, groups=groups,
+                                                                   reason=response.status.error_data, attempt=1))
+        else:
+            self.event_engine.trigger(events.HeartbeatSuccessEvent(channels=channels, groups=groups))
+
+
+class HeartbeatWaitEffect(Effect):
+    def __init__(self, pubnub_instance, event_engine_instance, invocation: invocations.HeartbeatWaitInvocation) -> None:
+        super().__init__(pubnub_instance, event_engine_instance, invocation)
+        self.heartbeat_interval = pubnub_instance.config.heartbeat_interval
+
+    def run(self):
+        if hasattr(self.pubnub, 'event_loop'):
+            self.stop_event = self.get_new_stop_event()
+            self.run_async(self.heartbeat_wait(self.heartbeat_interval, stop_event=self.stop_event))
+
+    async def heartbeat_wait(self, wait_time: int, stop_event):
+        try:
+            await asyncio.sleep(wait_time)
+            self.event_engine.trigger(events.HeartbeatTimesUpEvent())
+        except asyncio.CancelledError:
+            pass
+
+
+class HeartbeatLeaveEffect(Effect):
+    def run(self):
+        channels = self.invocation.channels
+        groups = self.invocation.groups
+        if hasattr(self.pubnub, 'event_loop'):
+            self.stop_event = self.get_new_stop_event()
+            self.run_async(self.leave(channels=channels, groups=groups, stop_event=self.stop_event))
+
+    async def leave(self, channels, groups, stop_event):
+        leave_request = Leave(self.pubnub).channels(channels).channel_groups(groups).cancellation_event(stop_event)
+        leave = await leave_request.future()
+
+        if leave.status.error:
+            self.logger.warning(f'Heartbeat failed: {leave.status.error_data.__dict__}')
+
+
+class HeartbeatDelayedEffect(Effect):
+    def __init__(self, pubnub_instance, event_engine_instance,
+                 invocation: Union[invocations.PNManageableInvocation, invocations.PNCancelInvocation]) -> None:
+        super().__init__(pubnub_instance, event_engine_instance, invocation)
```

Comment:
The same, let's think about improving it while working on a new reconnection policy